src/classes/actions-cache/upload.ts

```diff
@@ -0,0 +1,180 @@
+import * as core from '@actions/core';
+import fs from 'fs';
+import {HttpClient} from '@actions/http-client';
+import {TypedResponse} from '@actions/http-client/lib/interfaces';
+import {ReserveCacheError, ValidationError} from '@actions/cache';
+import {isSuccessStatusCode} from './http-responses';
+import {retryHttpClientResponse, retryTypedResponse} from './retry';
+import {getOctokit} from '@actions/github';
+import {retry as octokitRetry} from '@octokit/plugin-retry';
+import {createHttpClient, getCacheApiUrl} from './http-client';
+
+const uploadFile = async (
+  httpClient: HttpClient,
+  cacheId: number,
+  filePath: string,
+  fileSize: number
+): Promise<void> => {
+  if (fileSize <= 0) return;
+  const start = 0;
+  const end = fileSize - 1;
+  const contentRange = `bytes ${start}-${end}/*`;
+  core.debug(
+    `Uploading chunk of size ${
+      end - start + 1
+    } bytes at offset ${start} with content range: ${contentRange}`
+  );
+
+  const additionalHeaders = {
+    'Content-Type': 'application/octet-stream',
+    'Content-Range': contentRange
+  };
+
+  const resourceUrl = getCacheApiUrl(`caches/${cacheId.toString()}`);
+  const fd = fs.openSync(filePath, 'r');
+  try {
+    const uploadChunkResponse = await retryHttpClientResponse(
+      `uploadChunk (start: ${start}, end: ${end})`,
+      async () => {
+        const stream = fs
+          .createReadStream(filePath, {
+            fd,
+            start,
+            end,
+            autoClose: false
+          })
+          .on('error', error => {
+            throw new Error(
+              `Cache upload failed because file read failed with ${error.message}`
+            );
+          });
+        return httpClient.sendStream(
+          'PATCH',
+          resourceUrl,
+          stream,
+          additionalHeaders
+        );
+      }
+    );
+
+    if (!isSuccessStatusCode(uploadChunkResponse.message.statusCode)) {
+      throw new Error(
+        `Cache service responded with ${uploadChunkResponse.message.statusCode} during upload chunk.`
+      );
+    }
+  } finally {
+    fs.closeSync(fd);
+  }
+  return;
+};
+
+const resetCacheWithOctokit = async (cacheKey: string): Promise<void> => {
+  const token = core.getInput('repo-token');
+  const client = getOctokit(token, undefined, octokitRetry);
+  // TODO: better way to get repository?
+  const repo = process.env['GITHUB_REPOSITORY'];
+  core.debug(`remove cache "${cacheKey}"`);
+  try {
+    // TODO: replace with client.rest.
+    await client.request(
+      `DELETE /repos/${repo}/actions/caches?key=${cacheKey}`
+    );
+  } catch (error) {
+    if (error.status) {
```

Comment:
Shouldn't we check for specific status?

Reply:
There's no too much info about the possible status and since `octokit` is high-level wrapper i expect the thrown error object contains all the necessary info